File: src/main/java/com/leetcode/剑指 Offer 56 - I. 数组中数字出现的次数.java
/**
 * @author Bao Wenjie
 * @link https://leetcode-cn.com/problems/shu-zu-zhong-shu-zi-chu-xian-de-ci-shu-lcof/
 * @date 2021/10/14
 */

package com.leetcode;

class Solution56 {
    public int[] singleNumbers(int[] nums) {
        int ans = 0;
        for(int x : nums) {
            ans ^= x;
        }
        int index = 0;
        while(ans != 0) {
            if((ans & 0b1) != 0) {
                break;
            }
            ans >>= 1;
            index++;
        }
        int num1 = 0, num2 = 0;
        for(int num : nums) {
            if (((num >> index) & 0b1) == 0) {
                num1 ^= num;
            } else {
                num2 ^= num;
            }
        }
        int[] a = {num1, num2};
        return a;
    }
}

File: leetcode/Code/剑指 Offer 56 - I. 数组中数字出现的次数.cpp
/**
 * Author: Bao Wenjie
 * Date: 2021/3/22
 * Link: https://leetcode-cn.com/problems/shu-zu-zhong-shu-zi-chu-xian-de-ci-shu-lcof/
 */

#include <iostream>
#include <vector>

using namespace::std;


class Solution {
public:
    vector<int> singleNumbers(vector<int>& nums) {
    	int num = 0;
    	for(int x:nums)
    	{
    		num ^= x;
		}
		int gap = 1;
		for(int i = 0; i < 16; i++)
		{
			if(gap & num)
			{
				break;
			}
			gap <<= 1;
		}
		int num_ = 0;
		num = 0;
		for(int x:nums)
		{
			if(gap & x)
			{
				num ^= x;
			}
			else
			{
				num_ ^= x;
			}
		}
		vector<int> ans = {num, num_};
		return ans;
    }
};
